client/core/account.go

```diff
@@ -208,24 +211,51 @@ func (c *Core) AccountImport(pw []byte, acct *Account, bonds []*db.Bond) error {
 
 // UpdateCert attempts to connect to a server using a new TLS certificate. If
 // the connection is successful, then the cert in the database is updated.
+// However, it will attempt to reconnect to dex using the old cert if the new
+// connection fails.
 func (c *Core) UpdateCert(host string, cert []byte) error {
-	accountInfo, err := c.db.Account(host)
+	acct, err := c.db.Account(host)
 	if err != nil {
 		return err
 	}
 
-	accountInfo.Cert = cert
+	var connected bool
+	defer func(cert []byte) {
+		if connected {
+			return
+		}
+
+		// Attempt to reconnect using old cert.
+		acct.Cert = cert
+		c.connectAccount(acct)
+	}(acct.Cert)
+
+	// Remove previous dex connection first.
+	c.connMtx.RLock()
+	dc, found := c.conns[acct.Host]
+	c.connMtx.RUnlock()
+	if found {
+		c.disconnectDEX(dc)
+	}
 
-	_, connected := c.connectAccount(accountInfo)
-	if !connected {
-		return errors.New("failed to connect using new cert")
+	acct.Cert = cert
+	dc, err = c.connectDEX(acct)
+	if dc == nil {
+		return fmt.Errorf("failed to connect using new cert (will restore old connection): %v", err)
+	}
+	if err != nil {
+		dc.connMaster.Disconnect() // stop any retry loop.
+		return fmt.Errorf("failed to connect using new cert (will restore old connection): %v", err)
 	}
 
-	err = c.db.UpdateAccountInfo(accountInfo)
+	err = c.db.UpdateAccountInfo(acct)
 	if err != nil {
 		return fmt.Errorf("failed to update account info: %w", err)
 	}
 
+	connected = true
+	c.addDexConnection(dc)
```

Comment:
In other words, these lines are the danger zone where you've lost the `dexConnection` from the `Core.conns` map, and it may not come back at all, and if it does it come back, it was gone for a notable period.

Reply:
> In other words, these lines are the danger zone where you've lost the dexConnection from the Core.conns map

We can stop the previous conn and leave it in the map if this brief duration is going to be an issue.

Reply:
@chappjc, I believe your concern has been adequately handled in https://github.com/decred/dcrdex/pull/2019/commits/e776e8336885970a833233f970b2e12d962cdc3d, correct?

Reply:
Yes, thanks.

An unrelated/existing issue I think is what happens if you try to use `UpdateCert` on a host on which you are authorized (`authDEX` run).  Am I missing something or does this wipe out most of the important fields of the `dexConnection` when it is reconstructed freshly by `connectDEX`?  Have we just not tested doing a cert change on a functional DEX host?  I know the cert change function was added to deal with a DEX that has connectivity issues because of an expired certificate or something similar, but...  how was this supposed to work after it succeeds updating?

Reply:
> wipe out most of the important fields of

What `important` fields are you specifically referring to?

Reply:
Most of the `acct` (`dexAccount`) fields, and everything set by `authDEX`.
`trades` and everything set by `resolveActiveTrades`.

`connectDEX` makes a nerfed `dexConnection`, and `connectAccount` is a very narrowly purposed helper generally only used from `initialize` (pre-login).

This PR is fine though.  We can look at this in a follow-up.

Reply:
Ah, I see.

> We can look at this in a follow-up. 

Alright.

Reply:
We could also only allow updating unless there are no active orders, so nothing would break.  Similar to the logout checks.

Reply:
Even worse, every `trackedTrade` carries a reference to the previous `dexConnection`.  I think I'm convinced we should prevent a cert update unless the connection is down or their are no active trades.  And updating in the first case would still require a restart to get any trades moving again.

Reply:
> A more correct change would be to simply deny an update if the dexConnection is already connected

I agree.

Reply:
> Even worse, every trackedTrade carries a reference to the previous dexConnection. I think I'm convinced we should prevent a cert update unless the connection is down or there are no active trades. And updating in the first case would still require a restart to get any trades moving again.

The issues related to this seem to be a rabbit hole. However, we could do the following:

1. If we have an already connected dex, reject the update cert request. https://github.com/decred/dcrdex/pull/2019/commits/8ccc00a85b2f2e732611683bf9631e3004298066
2. If not connected, use a new `*dexConnection`(even if host is found in the conns map), initialize the dex connection(i.e `authDEX`), then load and resume active trades. This would mean `core.UpdateCert` will require the app's password so we can resume trades without requiring a login to trigger trade resumption for the newly connected DEX.  Any thoughts? WIP.

Reply:
> However, I'd like to first correct an unfortunate design pattern, which is having the "parent" types like dexConnection be fields of member types like trackedTrade. I've just completed a commit on a branch I can push shortly to illustrate.

Okay, sounds good.